### ISSUE TITLE
Enhancement: showing stacked tokens

### ISSUE THREAD

**Issue by colonelfazackerley** [andybalaam/rabbit-escape#399]
I am working on a way of having the tokens in a cell listed. Handy if there are several in a cell.

- pause
- click on a cell
- a list appears
- it goes after a seconds or so.

The same kind of thing should work in swing and droid.

(I will get back to the water polygons after)

**Comment by andybalaam** [comment]
Sounds good. I am also interested in stacked rabbits. I was thinking about slightly offsetting the images when there are multiple of the same one. That would give a general impression, and would be complementary to what you propose I think.

**Comment by colonelfazackerley** [comment]
I have it working for rabbits too. I have a few things left to do
- make it easier to share code with android
- positioning of the text

**Comment by colonelfazackerley** [comment]
![clicktips](https://cloud.githubusercontent.com/assets/9596719/14582114/1e4f047c-03f4-11e6-8631-343199b6a266.png)

**Comment by colonelfazackerley** [comment]
I just tried bold and blue, but it's a little hard to see with the stuff behind. i will do a semi-transparent rectangle behind the words.

**Comment by colonelfazackerley** [comment]
not bold, but blue. subtle semi transparent white behind the text. 

![clicktips2](https://cloud.githubusercontent.com/assets/9596719/14582913/0ddacf16-040a-11e6-85d6-0fcff0b163d0.png)

**Comment by andybalaam** [comment]
Thoughts:
* I'd prefer "2 x Explode" instead of "explode explode"
* I'd prefer centred on the square (vertically and horizontally)
* No need to say which way rabbits are facing
* Arguably no need to say what number a rabbit is

If we can make this attractive enough, it will be good to have it on pause, but I'm worried it will look messy no matter how much we try.

**Comment by andybalaam** [comment]
Maybe a long press on pause could give developer info.

**Comment by colonelfazackerley** [comment]
rabbit number is important. Elder rabbits acquire tokens first when the rabbits are stacked.

**Comment by colonelfazackerley** [comment]
I realised I had incidentally put in place most of the code to fix #80. So I fixed #80.

**Comment by colonelfazackerley** [comment]
In one of these levels 2 rabbits escape. In the other 1 escapes.

```
    j  
 r     
### ###
### #P#
### # #
O  b# #
#######
```

```
  r    
     j 
### ###
### #P#
### # #
O  b# #
#######
```

At this point I can see that this is the 1 rabbit dies scenario. Without rabbit index, I could not tell from this screenshot alone.
![order-is-important](https://cloud.githubusercontent.com/assets/9596719/14583158/44c9115c-0411-11e6-9d10-be58d8887222.png)

**Comment by colonelfazackerley** [comment]
I have done the centring and the counting (so it does not say "explode explode" anymore.

My example above also demonstrates the importance of facing.

**Comment by tttppp** [comment]
I can see how this would be useful, but it looks very "debug-mode" at the moment. I had a couple of ideas for similar things that looked cooler.

The Matrix "seeing the code" seems similar to this, and that would involve fading everything else out, and having a green text overlay.

Alternatively airforce pilots sometimes have an annotated display of targets (in films anyway!) This would probably need more arrows and circles drawing on, in addition to the text/abbreviations.

**Comment by tttppp** [comment]
![tmp_20115--the-matrix-the-matrix-23939329-1360-768-489222419](https://cloud.githubusercontent.com/assets/7073263/14585419/945b6ec6-046c-11e6-8e47-414164163814.jpg)
![tmp_20115-gug26bul178230214](https://cloud.githubusercontent.com/assets/7073263/14585428/c8dfd7cc-046c-11e6-99f3-9edcfa333492.jpg)

**Comment by tttppp** [comment]
Here's something I drew in a couple of minutes, I'm not sure how easy it would be to make:
![tmp_18395-2016-04-17 07 39 44-489222419](https://cloud.githubusercontent.com/assets/7073263/14585484/fb96d1b8-046f-11e6-9e31-adebf23b6403.png)

**Comment by colonelfazackerley** [comment]
I like that last one. If Balaam likes it too. I will put some effort in.

**Comment by colonelfazackerley** [comment]
What I have at the moment is annotation that appears when you click on a cell. I guess this should be an overlay that appears in pause mode, and disappears instantly when unpaused.

**Comment by tttppp** [comment]
Something I thought of while drawing that was stack order for tokens. Does that matter?

Another idea for an overlay is the text representation. Not sure how usable that would be. Also, where would the star points go?

**Comment by colonelfazackerley** [comment]
Token order does matter. The oldest token is picked up by the first rabbit to wander past. I changed it to this.
![overlay](https://cloud.githubusercontent.com/assets/9596719/14585693/99631cda-0475-11e6-91eb-05b3cebfdc4f.png)

Token order is here already. The older tokens are at the top of the list for a cell.

**Comment by colonelfazackerley** [comment]
Since then I made the text a less white and a stronger green. I darkened the grey overlay. It looks more like @tttppp's mockup and better.

**Comment by tttppp** [comment]
I like that mock up much more than the earlier ones. I'd suggest that if the quantity is 1 then we shouldn't show it. As a future improvement, for entrances and exits we could show the count of rabbits still to come/who've left.

**Comment by colonelfazackerley** [comment]
I was going to get rid of the 1's, that was a mistake.

I like the entrance/exit counter idea.

**Comment by colonelfazackerley** [comment]
I have "Exit\n3 saved" now. When there are multiple exits it is a bit wrong, as it is telling you how many have been saved from the world, not necessarily from that exit. However, I should not overcomplicate things.

**Comment by colonelfazackerley** [comment]
Same for multiple entrances. I am just giving the number of rabbits waiting. They wont all come in the same way.

**Comment by colonelfazackerley** [comment]
![overlay2](https://cloud.githubusercontent.com/assets/9596719/14585932/81b545b8-047f-11e6-80ca-1b8529663125.png)

**Comment by tttppp** [comment]
Did you change to 1-based indexing? If you can centre the exit and entrance text then I think this is pretty good.

**Comment by colonelfazackerley** [comment]
The text is centred. I guess you are telling me I need to centre each line.

**Comment by tttppp** [comment]
Ah, yes I see. I think it should be both centred in the cell and centre-aligned.

**Comment by andybalaam** [comment]
This is looking really cool! I agree with everything @tttppp said and am excited about this. Maybe it should just always look like this when paused.

**Comment by colonelfazackerley** [comment]
![overlay3](https://cloud.githubusercontent.com/assets/9596719/14590244/d5ba064e-04ed-11e6-9c98-69d9aa4cc49e.png)

done
- All annotations now appear when the game is paused
- Each line is individually centred.

feedback
- The oldest rabbit is 0: Does anyone object?
- Is this OK for now? Do we want circles and arrows instead of "[0]>" for rabbits?
- do we want token abbreviations (k instead of block)? this would help folk learn the abbreviations for use in level development

todo
- this breaks android.

**Comment by colonelfazackerley** [comment]
early opinions on the above questions are welcome.

![screenshot_20160418-214655](https://cloud.githubusercontent.com/assets/9596719/14619423/719a1fa6-05af-11e6-806e-fd333c915a38.png)

todo
 - rabbit indexing needs to restart for new levels.
- i need to force a redraw to show the overlay on android. it does not often come up.

**Comment by andybalaam** [comment]
> The oldest rabbit is 0: Does anyone object?

I don't.

> Is this OK for now? Do we want circles and arrows instead of "[0]>" for rabbits?

This looks cool.

> do we want token abbreviations

I would say no - learning abbreviations is quite extreme down the geeky scale.  Some day soon we'll have a level editor.

**Comment by tttppp** [comment]
I really like it. I feel like the casual player will be a little baffled by this screen, but are there actually any casual rabbit escape players?

As a future improvement we could have a setting that allowed the user to choose whether to show this on their pause screen (and default it to off). Now I've said this I'm not sure it's actually worth it.

Also would it be useful for displaying the water contents of each cell? Water changes every step, so maybe it's not worth it.

**Comment by colonelfazackerley** [comment]
I added water. it says "~340" to tell you the water content of a cell.

**Comment by colonelfazackerley** [comment]
i forced a redraw on android so the overlay shows when you pause.

todo: rabbit indexing needs to survive serialisation.

**Comment by colonelfazackerley** [comment]
I made a lot of progress with this. I just need to get the world to remember where it's up to with indexing. Otherwise, when the screen is turned on android, and an entrance births another bunny, it is number 1 again.

I went with the eldest rabbit being number 1. It made the save to state ever so slightly easier.

**Comment by colonelfazackerley** [comment]
I upgraded kubuntu. Now very little is working. Hope I can get some productivity back soon.

**Comment by colonelfazackerley** [comment]
fixed it, after like 6 hours of brick wall.
http://askubuntu.com/questions/760616/cannot-compile-java-following-upgrade-to-kubuntu-16-04/760890#760890

should make some more progress soon.

**Comment by andybalaam** [comment]
If we really think it will put people off we should not do it by default. We definitely do want casual players! However, I don't think it's actually off-putting, do you?

**Comment by colonelfazackerley** [comment]
I have played a bit, and I really like it.

Perhaps try it on 3 or so young playtesters to see how it goes with casual players*.

* Will that anger them?

**Comment by tttppp** [comment]
I don't think it will put them off, but I can certainly imagine casual players finishing most of the game without ever realising why you might want to consult this screen.

**Comment by andybalaam** [comment]
In the last screenshot you showed, it looks like the climbing rabbit is being shown as standing. Is that a bug?

**Comment by andybalaam** [comment]
I will try it on some troops. I think it should be in.

**Comment by colonelfazackerley** [comment]
> In the last screenshot you showed, it looks like the climbing rabbit is being shown as standing. Is that a bug?

yes it is. it's not my fault though. it's in the current release. i suggest treating that as a separate issue. i will go and file it now.

**Comment by colonelfazackerley** [comment]
implemented in #401
